### PR TITLE
[OM] Add support for Object instantiations and types.

### PR DESCRIPTION
**Pull request by mikeurbach**
This adds a new operation and type for Object instantiations according to the design in the rationale. Objects are intatiated by supplying a Class and actual parameters, and produce an SSA value of a new Class type. A simple round-trip test and error scenario tests are added.

### CODE REVIEW

**Comment by mikeurbach** [comment]
This is the third in a series of patches to begin a new Object Model dialect. The full patch series is here for reference: https://github.com/llvm/circt/compare/mikeurbach/om-dialect

File: lib/Dialect/OM/OMOps.cpp [hunk 1]
```diff
@@ -105,6 +105,46 @@ void circt::om::ClassOp::getAsmBlockArgumentNames(
...
+// ObjectOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult
+circt::om::ObjectOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+  // Get the containing ModuleOp.
+  auto moduleOp = getOperation()->getParentOfType<ModuleOp>();
+
+  // Verify the referred to ClassOp exists.
+  auto classDef = cast_or_null<ClassOp>(
```
Comment (uenoku) [inline lib/Dialect/OM/OMOps.cpp:118]: nit: can we use `dyn_cast_or_null`?

File: include/circt/Dialect/OM/OMOps.td [hunk 2]
```diff
@@ -58,4 +58,25 @@ def ClassFieldOp : OMOp<"class.field",
   }];
 }
 
+//===----------------------------------------------------------------------===//
+// Object instantiations
+//===----------------------------------------------------------------------===//
+
+def ObjectOp : OMOp<"object",
+    [DeclareOpInterfaceMethods<SymbolUserOpInterface>]> {
```
Comment (uenoku) [inline include/circt/Dialect/OM/OMOps.td:66]: Can we mark this Pure or do you have a plan to allow objects to have side-effects for some sense?

File: lib/Dialect/OM/OMOps.cpp [hunk 3]
```diff
@@ -105,6 +105,46 @@ void circt::om::ClassOp::getAsmBlockArgumentNames(
...
+
+  // Verify the actual parameter types match the formal parameter types.
+  for (size_t i = 0, e = actualTypes.size(); i < e; ++i) {
+    if (actualTypes[i] != formalTypes[i]) {
+      return emitOpError("actual parameter type (")
+             << actualTypes[i] << ") doesn't match formal parameter type ("
+             << formalTypes[i] << ')';
+    }
+  }
+
```
Comment (uenoku) [inline lib/Dialect/OM/OMOps.cpp:144]: Probably it would be nice to verify its result type is equal to class def.

**Review by uenoku — APPROVED** [review]
LGTM

File: include/circt/Dialect/OM/OMOps.td (hunk 2 above)
Comment (mikeurbach) [reply include/circt/Dialect/OM/OMOps.td:66]: Good question, I had thought a little bit about this... and in general I guess maybe we could have a world where this is not pure. But now that you bring it up, that's probably not a world we want to live in :) So I will add Pure, and we should maintain this restriction unless we really need to relax it.

**Pull request merged**